### PR TITLE
[Php71] Skip in if is_array() stmt on CountOnNullRector

### PR DESCRIPTION
**Pull request by samsonasik**
Given the following code:

```php
final class SkipInIfIsArray
{
    private $data;

    function countables()
    {
        if (is_array($this->data)) {
            echo count($this->data);
        }
    }
}
```

it currently produce:

```diff
         if (is_array($this->data)) {
-            echo count($this->data);
+            echo count((array) $this->data);
```

which should be skipped.

### CODE REVIEW

**Comment by samsonasik** [comment]
Fixed 🎉

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/skip_in_if_is_countable.php.inc [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    private $data;
+
+    function countables()
+    {
+        $this->data = rand(0, 1)
+            ? []
+            : null;
+
+        if (is_countable($this->data)) {
+            echo count($this->data);
```
Comment (TomasVotruba) [inline rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/skip_in_if_is_countable.php.inc:16]: This looks like simar bug to non empty string.

What is the phpstan type of this variable here?

Comment (samsonasik) [reply rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/skip_in_if_is_countable.php.inc:16]: the type is : `PHPStan\Type\Constant\ConstantArrayType` and `PHPStan\Type\ArrayType` by debug:

```php
dump($this->getType($funcCall->args[0]->value));
```

Comment (TomasVotruba) [reply rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/skip_in_if_is_countable.php.inc:16]: In that case, this condition should be ok.

The bug will be in type check then

Comment (TomasVotruba) [reply rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/skip_in_if_is_countable.php.inc:16]: This line should be probably extended with intersection type and similar check.

https://github.com/rectorphp/rector-src/pull/1387/files#diff-4e67b3812a474a7a969b9551d8bfa78c34512c53bc1f367cadafb774c97f7816R116

Line 116

Comment (samsonasik) [reply rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/skip_in_if_is_countable.php.inc:16]: Line 116? the isAlwaysIterableType() ? 

```php

        if ($this->isAlwaysIterableType($countedType)) {
            return null;
        }
```

Comment (TomasVotruba) [reply rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/skip_in_if_is_countable.php.inc:16]: Yes, now it checks only for union type, but it should include the intersection type too. Maybe the ComplexType node can be used also.

Comment (samsonasik) [reply rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/skip_in_if_is_countable.php.inc:16]: The check seems early on `CountableAnalyzer::isCastableArrayType()`, I moved check to it https://github.com/rectorphp/rector-src/pull/1387/commits/521bf6db7d6e6bdeabc8c78ac49c84480772bd50#diff-effb82d9a9df006478a5860e51fd141f60453d9f69b0d89bdef2b5bc0e1b3433

**Comment by samsonasik** [comment]
I added failing fixture where the data is filled by other method, and checked in `if is_array()` https://github.com/rectorphp/rector-src/pull/1387/commits/d661774ca376ff674c9f07de267e3b168b08d639

**Comment by samsonasik** [comment]
For code:

```php
final class SkipInIfIsArray2
{
    private $data;

    public function set(array $data)
    {
        $this->data = $data;
    }

    public function countables()
    {
        if (is_array($this->data)) {
            echo count($this->data);
        }
    }
}
```

It somehow got the following type:

```php
PHPStan\Type\ArrayType #14112
   keyType: PHPStan\Type\MixedType #14113
   |  isExplicitMixed: false
   |  subtractedType: null
   itemType: PHPStan\Type\MixedType #14109
   |  isExplicitMixed: false
   |  subtractedType: null
```

It will same if we define `@var mixed[]`:

```php
final class InIfIsArray2
{
    /** @var mixed[] */
    private $data;

    public function countables()
    {
        if (is_array($this->data)) {
            echo count($this->data);
        }
    }
}
```

which fine to be casted as never written by other method, but still strange as it already inside `is_array()` if.

**Comment by samsonasik** [comment]
@TomasVotruba I think the previous patch is the correct one by check inside condition `is_array()` or `is_countable()` https://github.com/rectorphp/rector-src/pull/1387/commits/a23e25a02557ad888a771cb020e3be2e28e4af29 as the code check in if condition, what do you think?

**Comment by TomasVotruba** [comment]
I'm on m way today, I'll check it tomorrow the latest in detail. Thanks for exploring the solutions.

**Comment by TomasVotruba** [comment]
Thanks for fixing the original issue.
Here is conflict of workaround for PHPStan type detection, that treats array docs as certains: https://phpstan.org/r/8217b74e-673d-43fe-a657-1ce9dae98d09

There should be possible count on `null` error, but it's ok. 
We'll have to find another way how to separate PHPStan false positive and `is_array()` valid check.	

I'm looking into it.

**Comment by TomasVotruba** [comment]
I tried to separate 3 cases:

```php
/**
 * @var mixed
 */
private $property;
```

Should not be property array

```php
/**
 * @var int[]
 */
private $property;
```

Should not be property array



```php
/**
 * @var int[]
 */
private $property = [];
```

Should be property array

It might need still some work for first 2 cases + is_array(). But better in next PR.

Could you review?

File: packages/NodeTypeResolver/TypeAnalyzer/ArrayTypeAnalyzer.php [hunk 2]
```diff
@@ -87,6 +96,41 @@ private function isIntersectionArrayType(Type $nodeType): bool
         return true;
     }
 
+    private function isPropertyFetchWithArrayDocblockWithoutDefault(Node $node): bool
+    {
+        if (! $node instanceof PropertyFetch && ! $node instanceof StaticPropertyFetch) {
+            return false;
+        }
```
Comment (samsonasik) [inline packages/NodeTypeResolver/TypeAnalyzer/ArrayTypeAnalyzer.php:104]: it is not needed as always PropertyFetch or StaticPropertyFetch

```suggestion
```

File: packages/NodeTypeResolver/TypeAnalyzer/ArrayTypeAnalyzer.php [hunk 3]
```diff
@@ -87,6 +96,41 @@ private function isIntersectionArrayType(Type $nodeType): bool
         return true;
     }
 
+    private function isPropertyFetchWithArrayDocblockWithoutDefault(Node $node): bool
```
Comment (samsonasik) [inline packages/NodeTypeResolver/TypeAnalyzer/ArrayTypeAnalyzer.php:100]:
```suggestion
    private function isPropertyFetchWithArrayDocblockWithoutDefault(PropertyFetch|StaticPropertyFetch $node): bool
```

File: packages/NodeTypeResolver/TypeAnalyzer/ArrayTypeAnalyzer.php [hunk 4]
```diff
@@ -97,10 +141,6 @@ private function isPropertyFetchWithArrayDefault(Node $node): bool
         }
```
Comment (samsonasik) [inline packages/NodeTypeResolver/TypeAnalyzer/ArrayTypeAnalyzer.php:142]: this check `if (! $node instanceof PropertyFetch && ! $node instanceof StaticPropertyFetch)` not needed as called inside `PropertyFetch` and `StaticPropertyFetch` check.

File: rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/count_on_class_method_stmts.php.inc [hunk 5]
```diff
@@ -1,31 +0,0 @@
...
-
-namespace Rector\Tests\Php71\Rector\FuncCall\CountOnNullRector\Fixture;
-
-use PhpParser\Node\Stmt\ClassMethod;
-
-final class CountOnClassMethodStmts
-{
-    public function run(ClassMethod $classMethod)
-    {
-        return count((array) $classMethod->stmts) === 5;
```
Comment (samsonasik) [inline rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/count_on_class_method_stmts.php.inc]: `$classMethod->stmts` can be null, so `(array)` cast is neeed.

File: packages/NodeTypeResolver/TypeAnalyzer/ArrayTypeAnalyzer.php (hunk 3 above)
Comment (TomasVotruba) [reply packages/NodeTypeResolver/TypeAnalyzer/ArrayTypeAnalyzer.php:100]: It's actually called from 2 places, where one is as you say, but the other takes always a `Node`:

PHPStan then reports:

image

File: rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/count_on_class_method_stmts.php.inc (hunk 5 above)
Comment (TomasVotruba) [reply rules-tests/Php71/Rector/FuncCall/CountOnNullRector/Fixture/count_on_class_method_stmts.php.inc]: I'll fix this one, thank you :+1:

**Comment by TomasVotruba** [comment]
Lets merge it as it is, to verify the fixes in practise. Thank you for fixes!

If you find another broken case, please make failing test fixture. I'll check it

**Pull request merged**